src/werkzeug/formparser.py

```diff
@@ -440,209 +357,79 @@ def get_part_charset(self, headers: Headers) -> str:
-    ) -> t.Iterator[t.Tuple[str, t.Tuple[str, t.Union[str, FileStorage]]]]:
-        """Generate ``('file', (name, val))`` and
-        ``('form', (name, val))`` parts.
-        """
-        in_memory = 0
-        guard_memory: bool
-        is_file: bool
-        container: t.Union[t.BinaryIO, t.List[bytes]]
-        _write: t.Callable[[bytes], t.Any]
-        headers: Headers
-        name: str
-        filename: str
-
-        for ellt, ell in self.parse_lines(file, boundary, content_length):
-            if ellt == _begin_file:
-                headers, name, filename = t.cast(t.Tuple[Headers, str, str], ell)
-                is_file = True
-                guard_memory = False
-                filename, container = self.start_file_streaming(
-                    filename, headers, content_length
-                )
-                _write = container.write
-
-            elif ellt == _begin_form:
-                headers, name = t.cast(t.Tuple[Headers, str], ell)
-                is_file = False
-                container = []
-                _write = container.append
-                guard_memory = self.max_form_memory_size is not None
-
-            elif ellt == _cont:
-                ell = t.cast(bytes, ell)
-                _write(ell)
-                # if we write into memory and there is a memory size limit we
-                # count the number of bytes in memory and raise an exception if
-                # there is too much data in memory.
-                if guard_memory:
-                    in_memory += len(ell)
-
-                    if in_memory > self.max_form_memory_size:  # type: ignore
-                        self.in_memory_threshold_reached(in_memory)
-
-            elif ellt == _end:
-                if is_file:
-                    container = t.cast(t.BinaryIO, container)
-                    container.seek(0)
-                    yield (
-                        "file",
-                        (name, FileStorage(container, filename, name, headers=headers)),
-                    )
-                else:
-                    part_charset = self.get_part_charset(headers)
-                    yield (
-                        "form",
-                        (name, b"".join(container).decode(part_charset, self.errors)),
-                    )
-
-    def parse(
-        self, file: t.BinaryIO, boundary: bytes, content_length: int
-    ) -> t.Tuple[MultiDict, MultiDict]:
-        form_stream, file_stream = tee(
-            self.parse_parts(file, boundary, content_length), 2
+            [None],
         )
-        form = (v for t, v in form_stream if t == "form")
-        files = (v for t, v in file_stream if t == "file")
-        return self.cls(form), self.cls(files)
+        parser = MultipartDecoder(boundary, self.max_form_memory_size)
+
+        forms = []
+        files = []
+
+        current_part: Union[Field, File]
+        for data in iterator:
+            parser.receive_data(data)
```

Comment:
Seems like this loop would be useful in the sansio class, as a higher level API that yields any available `key, value` pairs after sending data in. I'm imagining this method could then look something like this:

```python
decoder = MultipartDecoder()

while True:
    data = stream.read(self.buffer_size)
    
    if not data:
        break  # raise if decoder not completed?

    decoder.receive_data(data)

    for item in decoder:  # item = (key, value), iterator might be empty if waiting for more data
        if isinstance(item[1], str):
            fields.append(item)
        else:  # FileStorage
            files.append(item)

    # break if decoder completed?

Reply:
This would be a nicer API, however the file data needs to be streamed to disc as it is parsed (say for large files). This prevents it being yielded in a `key, value` pair as the disc access is IO and hence can not be SansIO.